### PR TITLE
Get and List mandates

### PR DESCRIPTION
**Pull request by mohammedmiah99**

### CODE REVIEW

File: src/TrueLayer/Mandates/IMandatesApi.cs [hunk 1]
```diff
@@ -1,9 +1,11 @@
 using System.Threading;
 using System.Threading.Tasks;
+using OneOf;
 
 namespace TrueLayer.Mandates
 {
     using TrueLayer.Mandates.Model;
+    using MandateDetailUnion = OneOf<Model.MandateDetail.AuthorizationRequiredMandateDetail, Model.MandateDetail.AuthorizingMandateDetail, Model.MandateDetail.AuthorizedMandateDetail, Model.MandateDetail.FailedMandateDetail, Model.MandateDetail.RevokedMandateDetail>;
```
Comment (Couper4) [inline src/TrueLayer/Mandates/IMandatesApi.cs:8]:
```suggestion
    using MandateDetailUnion = OneOf<
      Model.MandateDetail.AuthorizationRequiredMandateDetail,
      Model.MandateDetail.AuthorizingMandateDetail,
      Model.MandateDetail.AuthorizedMandateDetail,
      Model.MandateDetail.FailedMandateDetail,
      Model.MandateDetail.RevokedMandateDetail>;
```

File: src/TrueLayer/Mandates/MandatesApi.cs [hunk 2]
```diff
@@ -2,10 +2,12 @@
 using System.Threading;
 using System.Threading.Tasks;
 using TrueLayer.Auth;
+using OneOf;
 
 namespace TrueLayer.Mandates
 {
     using TrueLayer.Mandates.Model;
+    using MandateDetailUnion = OneOf<Model.MandateDetail.AuthorizationRequiredMandateDetail, Model.MandateDetail.AuthorizingMandateDetail, Model.MandateDetail.AuthorizedMandateDetail, Model.MandateDetail.FailedMandateDetail, Model.MandateDetail.RevokedMandateDetail>;
```
Comment (Couper4) [inline src/TrueLayer/Mandates/MandatesApi.cs:10]:
```suggestion
    using MandateDetailUnion = OneOf<
      Model.MandateDetail.AuthorizationRequiredMandateDetail,
      Model.MandateDetail.AuthorizingMandateDetail,
      Model.MandateDetail.AuthorizedMandateDetail,
      Model.MandateDetail.FailedMandateDetail,
      Model.MandateDetail.RevokedMandateDetail>;
```

File: src/TrueLayer/Mandates/MandatesApi.cs [hunk 3]
```diff
@@ -52,5 +54,48 @@ public async Task<ApiResponse<CreateMandateResponse>> CreateMandate(CreateMandat
...
+
+            ApiResponse<GetAuthTokenResponse> authResponse = await _auth.GetAuthToken(new GetAuthTokenRequest($"recurring_payments:{mandateType}"), cancellationToken);
+
+            if (!authResponse.IsSuccessful)
+            {
+                return new(authResponse.StatusCode, authResponse.TraceId);
+            }
+
+            return await _apiClient.GetAsync<MandateDetailUnion>(
+                new Uri(_baseUri, $"/v3/mandates/{mandateId}"),
```
Comment (Couper4) [inline src/TrueLayer/Mandates/MandatesApi.cs:71]: I think the resulting URI will be incorrect, it should be just `new Uri(_baseUri, "mandateId)`.


Having a look at how the [GetPayment method](https://github.com/1ststop/truelayer-dotnet/blob/dfc49f824b29d0bceb80c4fec43977ad70a541ab/src/TrueLayer/Payments/PaymentsApi.cs#L88) works might help here ;)

File: src/TrueLayer/Mandates/MandatesApi.cs [hunk 4]
```diff
@@ -52,5 +54,48 @@ public async Task<ApiResponse<CreateMandateResponse>> CreateMandate(CreateMandat
...
+        {
+            ApiResponse<GetAuthTokenResponse> authResponse = await _auth.GetAuthToken(new GetAuthTokenRequest($"recurring_payments:{mandateType}"), cancellationToken);
+
+            if (!authResponse.IsSuccessful)
+            {
+                return new(authResponse.StatusCode, authResponse.TraceId);
+            }
+
+            var queryParameters = System.Web.HttpUtility.ParseQueryString(string.Empty);
+            queryParameters["client_id"] = _options.ClientId;
```
Comment (Couper4) [inline src/TrueLayer/Mandates/MandatesApi.cs:88]: Why `client_id` is here?

I don't see it [mentioned in the docs](https://docs.truelayer.com/reference/list-mandate).

**Review by Couper4 — CHANGES_REQUESTED** [review]

File: test/TrueLayer.AcceptanceTests/MandatesTests.cs [hunk 5]
```diff
@@ -107,5 +108,39 @@ public async Task Can_create_mandate(CreateMandateRequest mandateRequest)
...
+            var createResponse = await _fixture.Client.Mandates.CreateMandate(
+                mandateRequest, idempotencyKey: Guid.NewGuid().ToString());
+            var mandateId = createResponse.Data!.Id;
+            // Act
+            var response = await _fixture.Client.Mandates.GetMandate(mandateId, MandateType.Sweeping);
+
+            // Assert
+            response.StatusCode.ShouldBe(HttpStatusCode.OK);
+            response.Data.AsT0.User!.Id.ShouldBe(createResponse.Data.User!.Id);
+            createResponse.StatusCode.ShouldBe(HttpStatusCode.Created);
```
Comment (Couper4) [inline test/TrueLayer.AcceptanceTests/MandatesTests.cs:127]:
```suggestion
            // Arrange
            var createResponse = await _fixture.Client.Mandates.CreateMandate(
                mandateRequest, idempotencyKey: Guid.NewGuid().ToString());
            createResponse.StatusCode.ShouldBe(HttpStatusCode.Created);
            var mandateId = createResponse.Data!.Id;
            
            // Act
            var response = await _fixture.Client.Mandates.GetMandate(mandateId, MandateType.Sweeping);

            // Assert
            response.StatusCode.ShouldBe(HttpStatusCode.OK);
            response.Data.AsT0.User!.Id.ShouldBe(createResponse.Data.User!.Id);
```

File: test/TrueLayer.AcceptanceTests/MandatesTests.cs [hunk 6]
```diff
@@ -107,5 +108,39 @@ public async Task Can_create_mandate(CreateMandateRequest mandateRequest)
...
+            // Arrange
+            var createResponse = await _fixture.Client.Mandates.CreateMandate(
+                mandateRequest, idempotencyKey: Guid.NewGuid().ToString());
+            // Act
+            var response = await _fixture.Client.Mandates.ListMandates(new ListMandatesQuery(createResponse.Data!.User.Id, null, 10), MandateType.Sweeping);
+
+            // Assert
+            response.StatusCode.ShouldBe(HttpStatusCode.OK);
+            response.Data!.Items.Count().ShouldBeLessThanOrEqualTo(10);
+            createResponse.StatusCode.ShouldBe(HttpStatusCode.Created);
```
Comment (Couper4) [inline test/TrueLayer.AcceptanceTests/MandatesTests.cs:143]:
```suggestion
            // Arrange
            var createResponse = await _fixture.Client.Mandates.CreateMandate(
                mandateRequest, idempotencyKey: Guid.NewGuid().ToString());
            createResponse.StatusCode.ShouldBe(HttpStatusCode.Created);
            
            // Act
            var response = await _fixture.Client.Mandates.ListMandates(new ListMandatesQuery(createResponse.Data!.User.Id, null, 10), MandateType.Sweeping);

            // Assert
            response.StatusCode.ShouldBe(HttpStatusCode.OK);
            response.Data!.Items.Count().ShouldBeLessThanOrEqualTo(10);
```

**Review by Couper4 — CHANGES_REQUESTED** [review]
I've left some last minor comments.

Can you please also merge the main branch into this one?

Other than that LGTM 👍

**Review by Couper4 — APPROVED** [review]

**Review by antoniovalentini — APPROVED** [review]

**Review by tl-antonio-valentini — APPROVED** [review]

**Comment by tl-antonio-valentini** [comment]
Builds seem to fail due to some secrets not being injected. They agreed with Mauro that if the unit tests pass it should be fine to merge.

**Pull request merged**